### ISSUE TITLE
three isses with SharpDX/MediaPlayer (MonoGame 3.4.0.456)

### ISSUE THREAD

**Issue by BenMakesGames** [mono/MonoGame#3884]
[edit] after playing around more, I discovered that the problems I was having locally are due to inconsistent application of Volume settings. This is a DIFFERENT sort of bug, and I uncovered an additional bug after further testing. I have edited the post accordingly.

---

I installed CrashReporter.NET into my project recently, and have been getting several reports related to MediaPlayer. I'm unable to reproduce these locally, but I've noticed some MediaPlayer issues of my own, and I have to wonder if they could be related.

First, the issues I have locally:

I am using MonoGame 3.4.0.456. My machine runs 64-bit Windows 7.

1. The first time a song is played, I receive "A first chance exception of type 'SharpDX.SharpDXException' occurred in SharpDX.dll". The game continues running as normal for me, however I wonder if this is the cause for some of the crashes I've had reported via CrashReporter.NET (below).

2. I set MediaPlayer.Volume every time a song is played, however the first time a particular song is played, this Volume is not respected; the song plays at 100% Volume. Subsequent plays of the same song respect the Volume I've set, but every time a song is played that has never been played before, the Volume is set back to full. (MonoGame seems to have some memory of Volume on a per-song basis, which fails to be applied the first time??) When a song loops - via MediaPlayer.IsRepeating = true - this seems to count as a second playthrough, and the previously-set Volume suddenly applies itself! I tried moving the code that sets MediaPlayer.Volume to both before and after the MediaPlayer.Play(...) call, but in both places, these symptoms remain.

3. Sometimes, a song simply refuses to play. This seems to happen randomly, to any song. There is no message in the console, or any other indication that something went wrong, and MediaPlayer.Status reports that the song is Playing.

4. I created a sound/song test menu within my game to facilitate easily testing the above.

Second: two MediaPlayer-related issues have been reported via CrashReporter.NET, but I have not been able to reproduce them locally. Here are the stack traces; I've truncated them at the point where it gets into my code, since I imagine that's not useful, but I can get the full stack traces including those, if it WOULD be helpful:

This report has only come from Windows 7 machines. I can't necessarily tell how many are UNIQUE reports, but I saw two different versions of .NET (4.0.30319.34209 and 4.0.30319.18063), so there are at least two unique versions (unless one player attempted to upgrade .NET to solve the issue?)

```
SharpDX.SharpDXException: HRESULT: [0x80004002], Module: [General], ApiCode: [E_NOINTERFACE/No such interface supported], Message: No such interface supported
 at SharpDX.Result.CheckError()
 at SharpDX.MediaFoundation.MediaFactory.GetService(ComObject unkObjectRef, Guid guidService, Guid riid, IntPtr& vObjectOut)
 at Microsoft.Xna.Framework.Media.MediaPlayer.GetVolumeObj(MediaSession session)
 at Microsoft.Xna.Framework.Media.MediaPlayer.PlatformPlaySong(Song song)
 at Microsoft.Xna.Framework.Media.MediaPlayer.Play(Song song)
```

AND:

This report has only come from machines running Vista. Similar to the first stacktrace, I can't be sure about the uniqueness of each report, but there are again two unique versions of .NET: 4.0.30319.34209 and 4.0.30319.18444.

```
SharpDX.SharpDXException: HRESULT: [0xC00D36E3], Module: [Unknown], ApiCode: [Unknown/Unknown], Message: Unknown
 at SharpDX.Result.CheckError()
 at SharpDX.MediaFoundation.MediaFactory.Startup(Int32 version, Int32 dwFlags)
 at SharpDX.MediaFoundation.MediaManager.Startup(Boolean useLightVersion)
 at Microsoft.Xna.Framework.Media.MediaManagerState.CheckStartup()
 at Microsoft.Xna.Framework.Media.Song.PlatformInitialize(String fileName)
 at Microsoft.Xna.Framework.Media.Song..ctor(String fileName)
 at Microsoft.Xna.Framework.Media.Song..ctor(String fileName, Int32 durationMS)
 at Microsoft.Xna.Framework.Content.SongReader.Read(ContentReader input, Song existingInstance)
 at Microsoft.Xna.Framework.Content.ContentTypeReader`1.Read(ContentReader input, Object existingInstance)
 at Microsoft.Xna.Framework.Content.ContentReader.InnerReadObject[T](T existingInstance)
 at Microsoft.Xna.Framework.Content.ContentReader.ReadObject[T](T existingInstance)
 at Microsoft.Xna.Framework.Content.ContentReader.ReadObject[T]()
 at Microsoft.Xna.Framework.Content.ContentReader.ReadAsset[T]()
 at Microsoft.Xna.Framework.Content.ContentManager.ReadAsset[T](String assetName, Action`1 recordDisposableObject)
 at Microsoft.Xna.Framework.Content.ContentManager.Load[T](String assetName)
```

If there's any other information I can provide, please let me know.

**Comment by mrhelmut** [comment]
Point 1 & 2 fixed by #4044.

Point 3 confirmed and still under work (#4090).

**Comment by MichaelDePiazzi** [comment]
FYI - PR #4090 has now been merged and resolves the remaining issues
